File: Dynamic Programming/Cutting Rod/python/cutting_rod.py
# bottom-up rod cutting dp method
def min_rod_cutting_cost(p, n):
    r = [0]*(n + 1)
    for i in range(1, n + 1):
        q = -1000000
        for j in range(1, i + 1):
            q = max(q, p[j - 1] + r[i - j])
        r[i] = q
    return r[n]

def extended_bottom_cut_rod(p, n):
    r, s = [0]*(n + 1), [0]*(n + 1)
    for i in range(1, n + 1):
        q = -1000000
        for j in range(1, i + 1):
            if q < (p[j - 1] + r[i - j]):
                q = (p[j - 1] + r[i - j])
                s[i] = j
        r[i] = q
    return (r, s)

def print_cut_rod(p, n):
    (r, s) = extended_bottom_cut_rod(p, n)
    while n > 0: # print steps
        print(s[n], end=" ")
        n -= s[n]

n = int(input())
p = [int(x) for x in input().split()]
print(min_rod_cutting_cost(p, n)) # minimum cut rod cost
print_cut_rod(p, n) # steps to reach minimum cost

File: Dynamic Programming/Cutting Rod/cpp/cutting_rod.cpp
#include <bits/stdc++.h>

using namespace std;

int extended_bottom_cut_rod(vector <int> p, int n, vector <int> &r, vector <int> &s) {
    for (int i = 1; i < n + 1; i++) {
        int q = -1000000;
        for (int j = 1; j < i + 1; j++) {
            if(q < (p[j - 1] + r[i - j])) {
                q = p[j - 1] + r[i - j];
                s[i] = j;
            }
        }
        r[i] = q;
    }
    return r[n];
}

void print_cut_rod(vector <int> p, int n) {
    vector <int> r(n + 1); vector <int> s(n + 1);
    int min_cut_cost = extended_bottom_cut_rod(p, n, r, s);
    cout << min_cut_cost << endl; // print minimum cost to cut rod
    // print steps of cutting rod 
    while(n > 0) {
        cout << s[n] << " ";
        n -= s[n];
    }
}

int main() {
    int n; cin >> n;
    vector <int> p(n);
    for (int i = 0; i < n; i++) {
        cin >> p[i];
    }
    print_cut_rod(p, n);
}
